Add validator score details to concepts (#43)
vegaprotocol/documentation: docs/concepts/vega-chain.md

hide_title: false
---

Vega uses Tendermint as a consensus layer to form a blockchain. The rest of the information here informs on how that blockchain and its relevant components is comprised. 

Read more: [How Vega bridges to Ethereum](/docs/concepts/vega-chain/#bridges-used-for-staking)

## Delegated proof of stake

> Vega runs on a delegated proof of stake blockchain. Participants -- validators and token-holders -- use their VEGA tokens to nominate validator nodes that run the network. Non-validator participants assign the voting rights of their VEGA tokens to endorse a validator's trustworthiness. 

Reviewer: suggestion (non-blocking) I wonder if this can do more to explain the slightly novel (to me) way that Vega does POS... that doesn't slash bad validator's stake but instead reduces the rewards they would have received from trading fees that epoch....
Token holders nominate validators they believe to be trust worthy, if a validator turns out to do things badly it (and it's delegates), forfeit staking rewards from trading fees that epoch. Then token holders may change their nomination, resulting in fewer rewards for un-trustworthy validators.

Reviewer 2: Agree with above suggestion."Vega is non slashing, i.e., there is no mechanism through which a token holder can lose a delegated token through a validator being punished; any measures to that end use different mechanism that will hurt a bad validators (and potentially their delegators) revenue, but not the delegated tokens.

Author: Non-slashing added!

> Vega runs on a delegated proof of stake blockchain. Participants -- validators and token-holders -- use their VEGA tokens to nominate validator nodes that run the network. Non-validator participants assign the voting rights of their VEGA tokens to endorse a validator's trustworthiness. 

Reviewer 2: Comment (non-blocking)This is slightly misleading about validator tokens. I would say "Participants (i.e., token-holders) use their Vega tokens....., whereas there is a minimum amount of tokens Validators are required to own and delegate to themselves.

> Vega runs on a delegated proof of stake blockchain. Participants -- validators and token-holders -- use their VEGA tokens to nominate validator nodes that run the network. Non-validator participants assign the voting rights of their VEGA tokens to endorse a validator's trustworthiness. 

Reviewer 2: In above comment, how about: In addition to being delectable to choose validators, tokens are also granting the token holder voting rights on governance decisions and network parameters. If a token is delegated, the voting rights stay with the token owner and are not transferred to the validator in question.

> Vega runs on a delegated proof of stake blockchain. Participants -- validators and token-holders -- use their VEGA tokens to nominate validator nodes that run the network. Non-validator participants assign the voting rights of their VEGA tokens to endorse a validator's trustworthiness. 

Author: I've reorganised this section a bit based on your comments here on dPoS, and hopefully it's more detailed and clearer.

Read more: [Validator nodes](/docs/concepts/vega-chain#validator-nodes)

**Participants who hold a balance of VEGA, the governance asset, can stake that asset on the network.** This is done by associating those tokens to a Vega key to use as stake, and then nominating one or more validators they trust to help secure the network. 

Everyone participating in keeping the network secure, robust and reliable, including nominators, is **rewarded** for keeping the network running. Not meeting the requirements of running the network can lead to penalties, such as **rewards being withheld**.

Read more: [Rewards](/docs/concepts/vega-chain#rewards)

Read more: [Penalties](/docs/concepts/vega-chain#penalties)

### VEGA token
Vega uses the VEGA ERC20 token for governance, which includes nominating validators to run nodes, and creating and voting on governance proposals.

A VEGA token (or fraction) can be either unassociated or associated with a Vega key:

* **Unassociated**: The tokenholder is free to do what they want with the token, but cannot nominate a validator with it

> * **Associated**: The token is locked in the staking smart contract and can be used to nominate a validator. It must be unassociated to withdraw it

Reviewer: suggest: Associated with a Vega key (using an ethereum smart contract) so that it can nominate a validator.

Author: I believe this is already covered by the line above, and I don't want to define associating by using the word associating :)

**All tokens can be used for staking and voting** on governance proposals. This includes tokens that are locked in the vesting contract. Tokens that are staked can be used to vote, and tokens used to vote can be staked.

Read more: [Governance of Vega](/docs/concepts/vega-protocol#governance)

:::info
A user's VEGA tokens must first be associated with a Vega key before they can be used for governance and staking.
:::

> ### Bridges used for staking

Reviewer: suggest this is something more like:
Vega listens to stake events from staking bridges (set by network param). At time of writing there are two, one for general use and one on the Vesting contract that holds tokens that are still locked according to a vesting schedule so that they can Stake and Govern while not being freely traded.

Author: I've incorporated the ideas from your staking bridge intro, let me know if you think it does the job

Because VEGA is an ERC20 token, all actions regarding staking are initiated on Ethereum, rather than on the Vega protocol. This allows VEGA to be staked to a Vega public key without any action on the Vega network, and without putting the tokens under the control of the Vega network.

> 

Reviewer 2: Comment (non-blocking): I find this a little misleading; the linking/unlinking a token to a Vega key is initiated by ethereal, the choice of validators is happening on Vega. Also might be useful to add something along the line of "All governance voting and change of nominations happen exclusively on the Vega chain, thus the only Ethereum gas fees that occur are the process of associating tokens and transferring the rewards.

Author: I've included your suggestions, and reorganised the section slightly to (hopefully) remove anything misleading about the eth/vega interaction

* Staking **unlocked tokens** is done using the staking bridge contract. The staking bridge contains functions enabling users to deposit, remove, and transfer stake by moving the governance tokens from their wallet to the staking bridge. 

> 

Reviewer 2: Comment (non-blocking) I couldn't call it moving the tokens, I interpret it more as notifying the staking bridge what tokens are associated to which key @C0deMunk33

Author: Updated to replace "moving tokens"!

hide_title: false
---

Vega uses Tendermint as a consensus layer to form a blockchain. The rest of the information here informs on how that blockchain and its relevant components is comprised. 

Read more: [How Vega bridges to Ethereum](/docs/concepts/vega-chain/#bridges-used-for-staking)

## Delegated proof of stake
Vega runs on a delegated proof of stake blockchain. 

> Validator nodes run the Vega network, and they deciding on the validity of the blocks containing the network's transactions and thus execute those transactions. The validators who run validator nodes are required to own a minimum amount of VEGA tokens that they delegate to themselves.

Reviewer 3: they *are* deciding?

Author: yes they are haha, thank you

> Validator nodes run the Vega network, and they deciding on the validity of the blocks containing the network's transactions and thus execute those transactions. The validators who run validator nodes are required to own a minimum amount of VEGA tokens that they delegate to themselves.

Author: ```suggestion
Validator nodes run the Vega network, and they decide on the validity of the blocks containing the network's transactions and thus execute those transactions. The validators who run validator nodes are required to own a minimum amount of VEGA tokens that they delegate to themselves.
```

All governance voting and validator nominations happen exclusively on the Vega chain. 

:::info
Ethereum gas fees are only incurred in the process of associating tokens to a Vega key and transferring rewards from a Vega key to an Ethereum address.
:::

The Vega protocol listens for stake events from staking bridges. Currently there are two bridges: one for staking unlocked, freely tradeable tokens, and one that connects to the vesting contract for locked, untradeable tokens. 

* Staking **unlocked tokens** uses the staking bridge contract. The staking bridge contains functions enabling users to deposit, remove, and transfer stake by notifying the staking bridge what tokens are associated to which Vega key. 

* When staking **locked tokens**, the Vega node interacts with the ERC20 vesting contract, which holds tokens that are locked per a vesting schedule, and provides the same utility as the staking bridge smart contract. This allows locked tokens to be used for staking and governance while not being freely tradeable. 

> Whether tokens are unlocked or locked, the bridge events make the Vega network of how many tokens a given party has associated and/or unassociated.

Reviewer 3: make the Vega network _know_ of how many

> Whether tokens are unlocked or locked, the bridge events make the Vega network of how many tokens a given party has associated and/or unassociated.

Author: ```suggestion
Whether tokens are unlocked or locked, the bridge events let the Vega network know of how many tokens a given party has associated and/or unassociated.
```

All events (including the above, plus stake per validator and others) are only registered after a certain number of block confirmations, as defined by the network parameter `blockchains.ethereumConfig`. 

:::note Further reading
**[Staking Bridge contracts](https://github.com/vegaprotocol/Staking_Bridge)** - on Vega's staking bridge GitHub repository.
:::

### Spam protection

> To avoid fragmentation or spam, there is a minimum delegateable stake that defines the smallest unit (fractions of) tokens that can be used for nomination, defined by the network parameter `validators.delegation.minAmount`. 

Reviewer 3: minAmount is not a very strong spam protection because the transaction will be processed, it will just be declined. There are better ones employed:
1) you need to have a balance of `spam.protection.delegation.min.tokens` to be able to submit a delegation transaction (I think that's currently 0.1 VEGA on mainnet) 
2) you can't run more than `spam.protection.max.delegations delegation` transactions per day.

> To avoid fragmentation or spam, there is a minimum delegateable stake that defines the smallest unit (fractions of) tokens that can be used for nomination, defined by the network parameter `validators.delegation.minAmount`. 

Author: ```suggestion
There are several spam protections enabled to protect the Vega network. 

- A participant who wants to submit a delegation (nomination) transaction, needs to have a balance of at least the minimum defined by the network parameter `spam.protection.delegation.min.tokens` to be able to submit the transaction. (Currently set as 0.1 VEGA)
- A participant cannot do more delegation transactions than the max set by `spam.protection.max.delegations` per day. (Currently set to 360.)
```

### Un-nominating validators
Participants can remove their nomination at the end of an epoch, or immediately. The un-nominated tokens will be restored back to the participant's associated token balance.

#### Un-nominate towards the end of the epoch
A participant can un-nominate towards the end of the current epoch, which means the stake is not used for the validator from the following epoch. 

The action is announced in the next available block of the same epoch, but the nominator keeps their nomination active until the last block of that epoch. At that point, the tokens are returned. The nominator cannot move those tokens before the epoch ends.

> 

Reviewer: Suggest including: this way a the token holder (and its nominated validator) is still entitled to the rewards from that epoch. using un-delgate now forfeits any rewards from that epoch

Reviewer 2: Add to above: 'un-delegate now or moving the tokens to a different Ethereum adress'

Author: I've added most of this, though in a slightly different form. 

Because the repercussions of 'un-nominate/delegate now' are described in that section, I left that as-is.

The participant will not receive any rewards from the validator in that epoch. The tokens are marked as available to the participant.

## Staking rewards & penalties

### Rewards
Validators and nominators both receive incentives for securing the network. The amount of those incentives, rewarded as VEGA, depends on factors including how much stake is nominated. 

**To be considered for staking rewards, a tokenholder must associate VEGA to a Vega key and nominate one or more validators.**

In each epoch, rewards are distributed among validators in proportion to the number of tokens they represent (i.e., their total stake). The total stake includes a validator's own stake and the tokens nominated to that validator. Of this reward, a fixed amount is distributed among the tokenholders the validator represents. Currently that is 88.3%, though validators can vote to change that value.

The reward scheme uses a linear reward curve - the reward per staked token is independent of the behaviour of other tokenholders. This holds for validators as well, with the exception that there is a maximum amount of stake an individual validator can take on. 

> 

Reviewer 2: However, to avoid validators getting too big, the reward a validator gets (and can distribute to its delegators) is capped, i.e., the reward per token is decreasing if a validator exceeds a maximum size.

Author: Added, thank you!

The action is announced in the next available block of the same epoch, but the nominator keeps their nomination active until the last block of that epoch. At that point, the tokens are returned. The nominator cannot move those tokens before the epoch ends.

#### Un-nominate now
A participant can choose to un-nominate at any time, and the action is executed immediately following the block it is announced in (within the same epoch). 

The participant will not receive any rewards from the validator in that epoch. The tokens are marked as available to the participant.

## Staking rewards & penalties

### Rewards
Validators and nominators both receive incentives for securing the network. The amount of those incentives, rewarded as VEGA, depends on factors including how much stake is nominated. 

**To be considered for staking rewards, a tokenholder must associate VEGA to a Vega key and nominate one or more validators.**

> In each epoch, rewards are distributed among validators in proportion to the number of tokens they represent (i.e., their total stake). The total stake includes a validator's own stake and the tokens nominated to that validator. Of this reward, a fixed amount is distributed among the tokenholders the validator represents. Currently that is 88.3%, though validators can vote to change that value.

Reviewer 3: the last statement is a bit inaccurate. this 0.883 is applied to the proportion of delegation by delegators, not on the reward amount that goes to the validator and its delegators. 

Lets say for example that validator 1 is due to get 1000 tokens for them and their delegators. 
And lets say that they have self stake of 3000 and delegation by delegators of 7000. 
That means that the delegators of validator1 will get 0.883 * 1000 * 7000/10000 = 618 which will be divided between the delegators to validator 1 based on their relative proportion of nomination in the validator. 
Validator 1 would get the rest 1000-618 = 382 tokens.

Author: I've cut that sentence for now but will reflect this better in another PR, thanks for explaining it!

> In each epoch, rewards are distributed among validators in proportion to the number of tokens they represent (i.e., their total stake). The total stake includes a validator's own stake and the tokens nominated to that validator. Of this reward, a fixed amount is distributed among the tokenholders the validator represents. Currently that is 88.3%, though validators can vote to change that value.

Author: ```suggestion
In each epoch, rewards are distributed among validators in proportion to the number of tokens they represent (i.e., their total stake). The total stake includes a validator's own stake and the tokens nominated to that validator. Of this reward, a fixed amount is distributed among the tokenholders the validator represents.
```

### Validator node performance
A validator node's performance is expressed through a validator score, and in the future, a performance score as well. 

The validator score is calculated for each epoch, based on how much stake a validator has as well as other factors including the total number of validators and the optimal stake. 

If, at the end of an epoch, a validator does not have sufficient stake self-nominated or has overall too much stake, then their validator score will be lowered, which can impact the rewards a validator and its' nominators receive. 

Below are the two factors that can lower a validator's score, and why. 

#### Not enough self-nominated stake
Self-nominated stake refers to the amount of VEGA a validator has staked to their own node.  The minimum stake amount required is set using the network parameter `reward.staking.delegation.minimumValidatorStake`. Not having enough self-nominated stake can have an impact on rewards. 

* **Network risk**: A validator who has not committed enough stake to meet the minimum is a risk to the network because they may not be invested in keeping the network running

> * **Validator score**: If a validator does not meet the `reward.staking.delegation.minimumValidatorStake`, then the validator's score is set to zero

Reviewer 3: that's not yet true in 0.47.6. If that was correct neither the validator nor parties who nominated them wouldn't get a reward which is still not the case. The only sanction is that the validator itself isn't rewarded, but it's score is what it is based on its total stake.

Author: Ah, I see what you mean, both can't be true at the same time. I've changed to make it less specific

> * **Validator score**: If a validator does not meet the `reward.staking.delegation.minimumValidatorStake`, then the validator's score is set to zero

Author: ```suggestion
* **Validator score**: If a validator does not meet the `reward.staking.delegation.minimumValidatorStake`, the validator's score is lowered, which can affect rewards
```
